ps3.py

```diff
@@ -0,0 +1,237 @@
+    """
+    Returns True if word is in the word_list and is entirely
+    composed of letters in the hand. Otherwise, returns False.
+    Does not mutate hand or word_list.
+    """
+    word = word.lower()
+    if "*" in word:
+        word = word.replace("*", f"[{VOWELS}]")
+        for word_g in word_list:
+            if re.fullmatch(word, word_g) is not None:
+                return True
+        else:
+            return False
+    elif word in word_list:
+        for letter in word:
+            if letter not in hand or word.count(letter) > hand.get(letter, 0):
+                return False
+        else:
+            return True
+    else:
+        return False
+
+
+def calculate_handlen(hand):
+    """
+    Returns the length (number of letters) in the current hand.
+    """
+    count_letter = 0
+    for letter in hand.keys():
+        count_letter += hand[letter]
+    return count_letter
+
+
+def play_hand(hand, word_list, count_points):
+    """
+    The hand finishes when there are no more unused letters.
+    The user can also finish playing the hand by inputing two
+    exclamation points (the string '!!') instead of a word.
+    """
+    while calculate_handlen(hand) != 0:
+        print("Current Hand:", end=" ")
+        display_hand(hand)
+        user_word = input("Enter word, or “!!” to indicate that you are finished:")
+        if is_valid_word(user_word, hand, word_list):
+            score_word = get_word_score(user_word, calculate_handlen(hand))
+            count_points += score_word
+            print(f"{user_word} earned {str(score_word)}."
+                  f" Total: {count_points} points")
+            print(" ")
+        elif user_word == "!!":
+            return count_points
+        else:
+            print("This is not valid word. Please choose another word")
+        hand = update_hand(hand, user_word)
+    else:
+        return count_points
+
+
+def substitute_hand(hand, letter):
+    """
+    Allow the user to replace all copies of one letter in the hand (chosen by user)
+    with a new letter chosen from the VOWELS and CONSONANTS at random. The new letter
+    should be different from user's choice, and should not be any of the letters
+    already in the hand.
+
+    If user provide a letter not in the hand, the hand should be the same.
+
+    Has no side effects: does not mutate hand.
+    """
+    if letter in hand:
+        new_hand = hand.copy()
+        if letter in VOWELS:
+            while True:
+                new_letter = random.choice(VOWELS)
+                if new_letter not in hand.keys():
+                    new_hand[new_letter] = new_hand.pop(letter)
+                    return new_hand
+        elif letter in CONSONANTS:
+            while True:
+                new_letter = random.choice(CONSONANTS)
+                if new_letter not in hand.keys():
+                    new_hand[new_letter] = new_hand.pop(letter)
+                    return new_hand
```

Comment:
We choose letter from vowels AND consonants. Vowels letter can be substituted with consonats and vise versa.